lib/tag/each.js

```diff
@@ -75,30 +75,47 @@ function _each(dom, parent, expr) {
 
     })
 
-    // mount new
+    // mount new / reorder
     var nodes = root.childNodes,
         prev_index = [].indexOf.call(nodes, prev)
 
-    arrDiff(items, rendered).map(function(item, i) {
-
-      var pos = items.indexOf(item)
-
-      if (!checksum && expr.key) item = mkitem(expr, item, pos)
-
-      var tag = new Tag({ tmpl: template }, {
-        before: nodes[prev_index + 1 + pos],
-        parent: parent,
-        root: root,
-        loop: true,
-        item: item
-      })
-
-      tags.splice(pos, 0, tag)
+    items.forEach(function (item, i) {
+      // start index search from position based on the current i:
+      var pos = items.indexOf(item, i)
+      var oldPos = rendered.indexOf(item, i)
+
+      // if not found, search backwards from current i position:
+      !~pos && (pos = items.lastIndexOf(item, i))
+      !~oldPos && (oldPos = rendered.lastIndexOf(item, i))
+
+      if (!~oldPos) {
+        // mount new
+        if (!checksum && expr.key) item = mkitem(expr, item, pos)
+
+        var tag = new Tag({ tmpl: template }, {
+          before: nodes[prev_index + 1 + pos],
+          parent: parent,
+          root: root,
+          loop: true,
+          item: item
+        })
+
+        tags.splice(pos, 0, tag)
+        return
+      }
 
+      if (pos !== oldPos) {
+        // reorder
+        root.insertBefore(nodes[prev_index + oldPos + 1], nodes[prev_index + pos + 1])
+        rendered.splice(pos, 0, rendered.splice(oldPos, 1)[0])
+        tags.splice(pos, 0, tags.splice(oldPos, 1)[0])
+        return
+      }
+      expr.pos && (tags[pos][expr.pos] = pos) // should update the index?
```

Comment:
My guess is that you should update the `item` object since the tag inherit's its values on [this line](https://github.com/pakastin/riotjs/blob/master/lib/tag/tag.js#L55)